tests/pytests/test.py

```diff
@@ -4022,6 +4022,28 @@ def prepare_env(env):
                    'ADDR', f'{password}localhost:{env.envRunner.shards[i].port}', 'MASTER']
     env.expect('SEARCH.CLUSTERSET', 'MYID', '0', 'RANGES', str(env.shardsCount), *shards).ok()
 
+@skip(cluster=False)
+def test_rq_job_without_topology(env:Env):
+    env.expect(debug_cmd(), 'PAUSE_TOPOLOGY_UPDATER').ok()
+    env.expect(debug_cmd(), 'CLEAR_PENDING_TOPOLOGY').ok()
+    workers = 5
+    env.expect(config_cmd(), 'SET', 'WORKERS', workers).ok()
+
+    # Verify that the `SHARD_CONNECTION_STATES` debug command is blocked when the topology is not set.
+    try:
+        con = env.getConnection()
+        with TimeLimit(2, 'Failed waiting (SUCCESS!)'):
+            con.execute_command(debug_cmd(), 'SHARD_CONNECTION_STATES')
+            env.assertTrue(False, message='Expected to fail')
+    except Exception as e:
+        env.assertContains('Failed waiting (SUCCESS!)', str(e))
+
+    # Now re-set the topology and call the debug command again
+    env.expect('SEARCH.CLUSTERREFRESH').ok()
+    # We should also see the effect of setting the number of workers
+    env.expect(debug_cmd(), 'SHARD_CONNECTION_STATES').equal([ANY, [ANY] * (workers + 1)] * env.shardsCount)
```

Comment:
Nice test.
Had we wanted to test the actual scenario (updating `WORKERS` with replica shard) - would that mimic the buggy behavior in OSS? Is the `SEARCH.CLUSTERREFRESH` command sets the topology in replicas?

Reply:
The entire test is the closest we can get to the faulty behavior. We make sure we don’t know the topology and then push a first job to the UV thread. The refresh command is the OSS version of `clusterSet`. There is no real reason to reproduce it with replicas beside this is how we discovered it